### ISSUE TITLE
Wrong result column type for requested conversion when reading JSONB_ARRAY nested in ROW in DB2

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13643]
This test doesn't work in Db2:

```java
JSONB j2 = create().fetchValue(row(jsonbArray(val(1), val(2)))).value1();
assertJSONEquals("[1,2]", j2.data());
```

There's an exception like this:

```
org.jooq.exception.DataAccessException: SQL [select json_array(cast(? as integer), cast(? as integer)) "nested.json_array" from SYSIBM.DUAL]; Error while reading field: row (json_array(1, 2)), at JDBC index: 1
	at org.jooq_3.16.7-SNAPSHOT.DB2.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:3095)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:670)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1464)
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1425)
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:238)
	at org.jooq.impl.Tools.fetchOne(Tools.java:2413)
	at org.jooq.impl.ResultQueryTrait.fetchOne(ResultQueryTrait.java:506)
	at org.jooq.impl.Tools.attach(Tools.java:1360)
	at org.jooq.impl.DefaultDSLContext.fetchOne(DefaultDSLContext.java:4598)
	at org.jooq.impl.DefaultDSLContext.lambda$fetchValue$47(DefaultDSLContext.java:4618)
	at org.jooq.impl.Tools.attach(Tools.java:1360)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4618)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4628)
	at org.jooq.test.all.testcases.MultisetTests.testNestedRowMappingWithJSON(MultisetTests.java:2405)
	at org.jooq.test.jOOQAbstractTest.testNestedRowMappingWithJSON(jOOQAbstractTest.java:4753)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.testcontainers.containers.FailureDetectingExternalResource$1.evaluate(FailureDetectingExternalResource.java:30)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.sql.SQLException: Error while reading field: row (json_array(1, 2)), at JDBC index: 1
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1588)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1525)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1484)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:143)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1449)
	... 47 more
Caused by: java.sql.SQLException: Error while reading field: "nested.json_array", at JDBC index: 1
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1588)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1525)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1484)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:143)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1567)
	... 51 more
Caused by: com.ibm.db2.jcc.am.SqlSyntaxErrorException: [jcc][t4][1092][11638][4.29.24] Invalid data conversion: Wrong result column type for requested conversion. ERRORCODE=-4461, SQLSTATE=42815
	at com.ibm.db2.jcc.am.b7.a(b7.java:810)
	at com.ibm.db2.jcc.am.b7.a(b7.java:66)
	at com.ibm.db2.jcc.am.b7.a(b7.java:133)
	at com.ibm.db2.jcc.am.bi.Q(bi.java:1939)
	at com.ibm.db2.jcc.am.ResultSet.getBytes(ResultSet.java:1270)
	at org.jooq.tools.jdbc.DefaultResultSet.getBytes(DefaultResultSet.java:170)
	at org.jooq.tools.jdbc.DefaultResultSet.getBytes(DefaultResultSet.java:170)
	at org.jooq.tools.jdbc.DefaultResultSet.getBytes(DefaultResultSet.java:170)
	at org.jooq.impl.CursorImpl$CursorResultSet.getBytes(CursorImpl.java:547)
	at org.jooq.impl.DefaultBinding$DefaultBytesBinding.get0(DefaultBinding.java:2114)
	at org.jooq.impl.DefaultBinding$DefaultJSONBBinding.get0(DefaultBinding.java:5384)
	at org.jooq.impl.DefaultBinding$DefaultJSONBBinding.get0(DefaultBinding.java:5327)
	at org.jooq.impl.DefaultBinding$InternalBinding.get(DefaultBinding.java:1016)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1578)
	... 55 more
```

The workaround is to pass an explicit `RETURNING JSONB` clause, but it shouldn't be necessary. jOOQ should generate it automatically.

**Comment by lukaseder** [comment]
See also: https://github.com/jOOQ/jOOQ/issues/10434

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.7 (https://github.com/jOOQ/jOOQ/issues/13644), and 3.15.11 (https://github.com/jOOQ/jOOQ/issues/13645)

**Comment by lukaseder** [comment]
This also affected `JSONB_OBJECT()`, which is now also fixed.
